test/EndToEnd/Ledger.hs

```diff
@@ -208,3 +226,250 @@ unauthorisedBidderTest = mkAssertion $ do
+
+  terms <- liftIO $ do
+    dynamicState <- constructTermsDynamic seller utxoRef
+    configToAuctionTerms config dynamicState
+
+  assertNFTNumEquals seller 1
+
+  announceAuction terms
+
+  withActor buyer1 $ mkDeposit terms
+  withActor buyer2 $ mkDeposit terms
+
+  assertUTxOsInScriptEquals Deposit terms 2
+
+  waitUntil $ biddingStart terms
+  actorsPkh <- liftIO $ getActorsPubKeyHash [buyer1, buyer2]
+  startBidding terms (ApprovedBidders actorsPkh)
+
+  assertNFTNumEquals seller 0
+
+  withActor buyer1 $ newBid terms $ startingBid terms
+  withActor buyer2 $ newBid terms $ startingBid terms + minimumBidIncrement terms
+
+  waitUntil $ voucherExpiry terms
+
+  result <- try $ sellerClaimDepositFor terms (head actorsPkh)
+
+  case result of
+    Left (_ :: SomeException) -> assertUTxOsInScriptEquals Deposit terms 2
+    Right _ -> fail "Seller should not be able to claim deposit from losing bidder"
+
+bidderBuysWithDepositTest :: Assertion
+bidderBuysWithDepositTest = mkAssertion $ do
+  let seller = Alice
+      buyer1 = Bob
+      buyer2 = Carol
+
+  mapM_ (initWallet 100_000_000) [seller, buyer1, buyer2]
+
+  nftTx <- mintOneTestNFT
+  let utxoRef = mkTxIn nftTx 0
+
+  terms <- liftIO $ do
+    dynamicState <- constructTermsDynamic seller utxoRef
+    configToAuctionTerms config dynamicState
+
+  assertNFTNumEquals seller 1
+
+  announceAuction terms
+
+  withActor buyer1 $ mkDeposit terms
+  withActor buyer2 $ mkDeposit terms
+
+  assertUTxOsInScriptEquals Deposit terms 2
+
+  waitUntil $ biddingStart terms
+  actorsPkh <- liftIO $ getActorsPubKeyHash [buyer1, buyer2]
+  startBidding terms (ApprovedBidders actorsPkh)
+
+  assertNFTNumEquals seller 0
+
+  withActor buyer1 $ newBid terms $ startingBid terms
+  withActor buyer2 $ newBid terms $ startingBid terms + minimumBidIncrement terms
+
+  waitUntil $ biddingEnd terms
+  withActor buyer2 $ bidderBuys terms
+
+  assertNFTNumEquals seller 0
+  assertNFTNumEquals buyer1 0
+  assertNFTNumEquals buyer2 1
+  assertUTxOsInScriptEquals Deposit terms 1
+
+  waitUntil $ cleanup terms
+  cleanupTx terms
+
+cleanupDepositTest :: Assertion
+cleanupDepositTest = mkAssertion $ do
+  let seller = Alice
+      buyer1 = Bob
+      buyer2 = Carol
+
+  mapM_ (initWallet 100_000_000) [seller, buyer1, buyer2]
+
+  nftTx <- mintOneTestNFT
+  let utxoRef = mkTxIn nftTx 0
+
+  terms <- liftIO $ do
+    dynamicState <- constructTermsDynamic seller utxoRef
+    configToAuctionTerms config dynamicState
+
+  assertNFTNumEquals seller 1
+
+  announceAuction terms
+
+  withActor buyer1 $ mkDeposit terms
+  withActor buyer2 $ mkDeposit terms
+
+  assertUTxOsInScriptEquals Deposit terms 2
+
+  waitUntil $ biddingStart terms
+  actorsPkh <- liftIO $ getActorsPubKeyHash [buyer1, buyer2]
+  startBidding terms (ApprovedBidders actorsPkh)
+
+  assertNFTNumEquals seller 0
+
+  withActor buyer1 $ newBid terms $ startingBid terms
+  withActor buyer2 $ newBid terms $ startingBid terms + minimumBidIncrement terms
```

Comment:
Actually we can separate all beginig and rename them to `losingBidder` and `winningBidder`, cuz that case seems to be reused everywhere.

Reply:
could we do this as part of #267? It's worth thinking of how to avoid all this repetition in the tests (maybe we could test the cli directly? and use stuff like prepare-for-demo)